ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -224,26 +230,48 @@ public PullQueryResult execute(
           .getMaterialization(queryId, contextStacker)
           .orElseThrow(() -> notMaterializedException(getSourceName(analysis)));
 
-      final Struct key = asKeyStruct(whereInfo.keyBound, query.getPhysicalSchema());
+      final List<Optional<KsqlNode>> sourceNodes = new ArrayList<>();
+      final List<List<?>> tableRows = new ArrayList<>();
+      final List<LogicalSchema> schemas = new ArrayList<>();
+      final List<Future<PullQueryResult>> futures = new ArrayList<>();
+      final List<List<Struct>> keysByLocation = mat.locator().groupByLocation(
```

Comment:
Why don't you return besides they keys also the active, standby per group? This way you wouldn't need to do `locate` twice basically.

Reply:
Actually, I think currently the routing is wrong and you have to return the <active,standby> per group

Reply:
Yeah, I had thought that a given set of partitions were grouped together at each active and standby, but I think you're right this isn't the case.  I can fairly easily change this to be `keysByPartition` rather than host, though this isn't much better than just forwarding by individual key, unless the number of keys fetched is very large.  I'm going to think about if this can be easily reworked.

Reply:
I reworked this so that it now groups key together only if they share the same list of nodes, including active and all standbys.  Most of the time if there's 1 or 2 standbys and lots of keys fetched, this will hopefully reduce unnecessary calls.